include/common/bitdepth.rs

```diff
@@ -113,6 +115,56 @@ impl BPC {
         debug_assert!(n % scale == 0.into());
         n / scale
     }
+
+    /// Converts a value in bytes to a value in coef units.
+    ///
+    /// `T` is generally meant to be `usize` or `isize`.
+    pub fn coefstride<T>(&self, n: T) -> T
+    where
+        T: Copy + Eq + From<u8> + Div<Output = T> + Rem<Output = T>,
+    {
+        let scale = (self.bitdepth() / 4).into();
+        debug_assert!(n % scale == 0.into());
+        n / scale
+    }
+}
+
+/// # Safety
+///
+/// Caller is responsible for ensuring that the target type is valid to
+/// construct from an arbitrary byte pattern.
+unsafe fn cast_byte_slice<T>(bytes: &[u8]) -> &[T] {
+    let size = mem::size_of::<T>();
+
+    // Check that the number of elements is a multiple of the new element
+    // size and that the alignment is correct for the new element type.
+    debug_assert!(bytes.len() % size == 0);
+    assert!(bytes.as_ptr() as usize % mem::align_of::<T>() == 0);
+
+    let len = bytes.len() / size;
+
+    // SAFETY: We've checked that alignment and the number of elements is
+    // correct, caller is responsible for ensuring target type is valid.
+    unsafe { slice::from_raw_parts(bytes.as_ptr().cast(), len) }
+}
+
+/// # Safety
+///
+/// Caller is responsible for ensuring that the target type is valid to
+/// construct from an arbitrary byte pattern.
+unsafe fn cast_byte_slice_mut<T>(bytes: &mut [u8]) -> &mut [T] {
+    let size = mem::size_of::<T>();
+
+    // Check that the number of elements is a multiple of the new element
+    // size and that the alignment is correct for the new element type.
+    debug_assert!(bytes.len() % size == 0);
+    assert!(bytes.as_ptr() as usize % mem::align_of::<T>() == 0);
+
+    let len = bytes.len() / size;
+
+    // SAFETY: We've checked that alignment and the number of elements is
+    // correct, caller is responsible for ensuring target type is valid.
+    unsafe { slice::from_raw_parts_mut(bytes.as_mut_ptr().cast(), len) }
 }
```

Comment:
Can we replace these with [`zerocopy`](https://docs.rs/zerocopy)'s [`FromBytes::slice_from`](https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html#method.slice_from) and [`FromBytes::slice_from_mut`](https://docs.rs/zerocopy/latest/zerocopy/trait.FromBytes.html#method.slice_from_mut)?  For the 4 functions ([pixel, coef] x [&, &mut]), we can just use `FromBytes::slice_from(_mut)(bytes).unwrap()`.

This way `zerocopy` can do all of the safety checking for us here, and it does it with traits, which is a bit safer.  It's also written by google, so it should be fine as a dependency.  And I checked that these functions are zero cost abstractions.  We should also store the coef buffers as `i16`s instead of `u8`s, since they're either `i16` or `i32`, and this way the cast to `i16` is a no-op.  With our approach, we'd have to write `cast_i16_slice(_mut)` functions for that, but with `zerocopy`, we get that generically for free.

If you want, I can put this up as a separate PR.

Reply:
I think using zerocopy for the slice transmutes makes sense, I've added that. I don't think making the buffer `i16`s makes sense though. The cast should already be functionally a no-op, or at least very cheap, and I'd like to stick to the pattern of having byte buffers that either get cast to `Pixel` or `Coef` buffers.

Reply:
The cast to a larger type has a branch and panicking code, though it's still quite cheap.  The bpc8 casts are no-ops, though, when using the generic code if the target type is the same.  In general, though, I think it's clearer if we had a `type MinPixel = BitDepth8::Pixel` and `type MinCoef = BitDepth8::Coef` that we used instead of `u8` for these buffers.  Then it's clear that it's pixels or coefs vs. just some other `u8`s.

Reply:
Maybe we can make the definition of `Dyn{Pixel,Coef}` that.

Reply:
Could we just open an issue to do this later at some point?

Reply:
Sure, open an issue to look into it. It's probably only worth doing after profiling if we find that the cast is causing performance issues, but can't hurt to track the potential improvement.

Reply:
* Tracked in #932.